### ISSUE TITLE
TED-LIUM release 2 recipe 

### ISSUE THREAD

**Issue by vijayaditya** [kaldi-asr/kaldi#768]
User vince62s has updated the TED-LIUM recipe to use the TED-LIUM 2 release of data, which almost doubles the data. Obviously we want to add these changes to the recipe. However we would like to preserve the old recipe as it was a widely used benchmark. This gives us a nice opportunity to write a new TED-LIUM recipe called egs/tedlium/s5_r2 which uses new release 2 data.

This work involves going through the old recipe and selecting only the essential stages which build the best system(s). Modifying the existing scripts to use release 2 data. vince62s has already accomplished this but we would need to review these scripts and adapt them for use in the new scripts being written.

I would classify this as a mid-level expertise project and it would be an interesting exercise for users who would like to start contributing at the system building level. The biggest advantage here is that TED-LIUM is a publicly and freely available database so anyone can contribute without issues.

Further if you have issues of computational infra do not worry as we at CLSP can provide you access to our cluster.

More info:  (https://groups.google.com/d/topic/kaldi-help/VXn90AdDyy0/discussion)

**Comment by pranavj** [comment]
Where are his scripts? Didnt see them in egs/tedlium

**Comment by vijayaditya** [comment]
@pranavj The updated recipe has not pushed yet. They reside in vince's repo at 
https://github.com/vince62s/kaldi/tree/tedlium-r2/egs/tedlium

**Comment by vijayaditya** [comment]
For the sake of clarity I am writing down the changes required in detail

1. Create a new directory tedlium/s5_r2 with a new recipe which uses tedlium release 2 data.

2. Make the necessary changes for modifying data prep scripts (
 Check the changes made by vince62s [here](https://github.com/vince62s/kaldi/tree/tedlium-r2/egs/tedlium/s5_r2) for adding the release-2 data to Ted-LIUM recipe.

3. Review the old scripts in the tedlium/s5 and figure our which GMM, nnet1 and nnet3 systems have the best numbers (Nnet3 recipes are still being reviewed and will soon be added to the master, see PR #644 ) and adapt these recipes to use the release-2 data.

We would like to eliminate any systems which are not state-of-art in the new set of scripts. e.g. we would no longer support [nnet2 scripts](https://github.com/kaldi-asr/kaldi/tree/master/egs/tedlium/s5/local/online) in this recipe.

**Comment by vijayaditya** [comment]
Nnet3 scripts necessary for this issue have been added to chain branch. See #783.

**Comment by vince62s** [comment]
@vijayaditya Just for clarification. the WER we see for nnet3 seem higher (ie not as good as) the best nnet2 system if I am not mistaken. is your plan to stick to nnet3 = SOA ?

**Comment by vijayaditya** [comment]
I have just seen differences in performance I consider minor between these
two toolkits.

The optimization algorithm was slightly modified in nnet3, as the new
toolkit does gradient clipping based on change of parameters in the entire
model rather than per individual layer, which might be one of the reasons
for this difference in the results.

Nnet3 has implementations for RNNs and it is convenient to compare models
from the same toolkit, so I don't think we would be investing time in nnet2
anymore.

--Vijay

**Comment by vince62s** [comment]
Just to be sure what we are comparing. nnet2 best score 10.9% after 4 epochs, versus nnet3 12.1% after 4 epochs, all based on relaese 1. Am I correct ?
I fully understand the nnet3 direction, but just to be sure where performance stands before moving myself. thanks.

**Comment by jtrmal** [comment]
Perhaps we see the influence of the improved/modified regularization?
Y.

**Comment by danpovey** [comment]
It looks like the reason for the difference is that the nnet3 cross-entropy
TDNN recipe is quite a bit behind the corresponding nnet2 TDNN recipe.  The
configuration is quite different and the nnet3 one seems to have some
problems.

For a start, in the nnet3 recipe in egs/tedlium/s5/local/nnet3/run_tdnn.sh,
results are given for 2 epochs and 4 epochs; the 4-epoch run has better
results but the script uses 2 epochs.

Also, the configuration does not look very good.
It's like this:

   --relu-dim 500 \

    --splice-indexes "-1,0,1 -1,0,1,2 -3,0,3 -3,0,3 -3,0,3 -6,-3,0" \

while the nnet2 one is p-norm with

    --pnorm-input-dim 3500 \

    --pnorm-output-dim 350 \

--splice-indexes "layer0/-2:-1:0:1:2 layer1/-1:2 layer3/-3:3 layer4/-7:2" \

One problem with the nnet3 splice indexes is that they end with too many
indexes; they should end with "0".  [in the nnet2 setup, the final "0" is
implicit].  By ending with "-6,-3,0", the final dimension is too high, at
1500, so too many parameters are in the final layer (1500 x 5000 or
whatever the num-leaves is).  It would be better to increase the relu dim,
e.g. to 750, and have the splice indexes more like the Switchboard setup
which is

 --splice-indexes "-2,-1,0,1,2 -1,2 -3,3 -7,2 0"  \

Vijay, can you please make sure someone looks into this?


Dan

**Comment by galv** [comment]
I can do begin an experiment within an hour using Dan's suggested configs. I haven't deleted my TEDLIUM setup.

I wrote the original nnet3 recipe. I was thinking only of chain-vs-nnet3 results at the time, so I did not realize the nnet3 model was worse than the nnet2 model at the time. The splice indices you see for nnet3 look like a hold-over from the chain model.

**Comment by danpovey** [comment]
cool.

**Comment by galv** [comment]
Running now. Bear in mind it will be - I think - ~2 days. Here are the relevant configs:

```
  python steps/nnet3/tdnn/make_configs.py  \
    --feat-dir data/${train_set}_hires \
    --ivector-dir exp/nnet3/ivectors_${train_set} \
    --ali-dir $ali_dir \
    --relu-dim 768 \
    --splice-indexes "-2,-1,0,1,2 -1,2 -3,3 -7,2 0"  \
    --use-presoftmax-prior-scale true \
   $dir/configs || exit 1;

  steps/nnet3/train_dnn.py --stage=$train_stage \
    --cmd="$decode_cmd" \
    --feat.online-ivector-dir exp/nnet3/ivectors_${train_set} \
    --feat.cmvn-opts="--norm-means=false --norm-vars=false" \
    --trainer.num-epochs 4 \
    --trainer.optimization.num-jobs-initial 3 \
    --trainer.optimization.num-jobs-final 8 \
    --trainer.optimization.initial-effective-lrate 0.0015 \
    --trainer.optimization.final-effective-lrate 0.00015 \
    --egs.dir "$common_egs_dir" \
    --cleanup.remove-egs $remove_egs \
    --cleanup.preserve-model-interval 20 \
    --feat-dir=data/${train_set}_hires \
    --ali-dir $ali_dir \
    --lang data/lang \
    --reporting.email="$reporting_email" \
    --dir=$dir  || exit 1;
```

**Comment by vince62s** [comment]
@galv any update on you run ? thanks.

**Comment by galv** [comment]
I've been a little pressed for time the past two days with travelling and then moving out a space. Sorry for the period of silence.

Here are my results form the latest run. I want to interpret these but I need to run now:


```
%WER 15.1 | 507 17792 | 88.0 8.9 3.1 3.1 15.1 89.0 | -0.073 | exp/nnet3/tdnn_new_splice/decode_dev/score_10_0.0/ctm.filt.filt.sys
%WER 13.6 | 507 17792 | 89.1 7.8 3.1 2.7 13.6 87.0 | -0.164 | exp/nnet3/tdnn_new_splice/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 13.6 | 1155 27512 | 88.7 8.5 2.8 2.3 13.6 83.5 | -0.052 | exp/nnet3/tdnn_new_splice/decode_test/score_10_0.0/ctm.filt.filt.sys
%WER 12.2 | 1155 27512 | 89.8 7.5 2.7 2.0 12.2 80.0 | -0.114 | exp/nnet3/tdnn_new_splice/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
```

**Comment by danpovey** [comment]
@galv, what is the file location for these results?  And how do they compare with your previous results?

**Comment by vince62s** [comment]
OK, I just created a PR since this issue was still pending.
I am currently running various results with Nnet2 and Nnet3.
I will do Chain later on if I get the time.

**Comment by vince62s** [comment]
@vijayaditya I think you can close this issue.

**Comment by vijayaditya** [comment]
Addressed in #905
